libdnf/transaction/Swdb.cpp

```diff
@@ -55,22 +57,52 @@ Swdb::Swdb(const std::string &path)
   , autoClose(true)
 {
     if (path == ":memory:") {
-        // writing to an in-memory database
-        conn = std::make_shared< SQLite3 >(path);
+        // connect to an in-memory database as requested
+        conn = std::make_shared<SQLite3>(path);
         Transformer::createDatabase(conn);
-    } else if (!pathExists(path.c_str())) {
-        // writing to a file that doesn't exist and must be created
-
-        // extract persistdir from path - "/var/lib/dnf/"
-        auto found = path.find_last_of("/");
-
-        Transformer transformer(path.substr(0, found), path);
-        transformer.transform();
-
-        conn = std::make_shared< SQLite3 >(path);
+    } else if (pathExists(path.c_str())) {
+        try {
+            // connect to an existing database
+            conn = std::make_shared<SQLite3>(path);
+            // try to execute a select to detect if the database is not read-only
+            conn->exec("BEGIN; UPDATE config SET value='test' WHERE key='test'; ROLLBACK;");
```

Comment:
Interesting test :)

Reply:
:slightly_smiling_face: 

The comment above is outdated (it no longer is a select).

Yeah, it's obviously not great, but I actually don't understand it (anymore).

As a non-root you're only doing selects, right, so why the update?

Also, you've said you need to write to the file because of some lock even when doing a select? Is the file writeable by users in the default system setup then? Because otherwise you wouldn't be able to do the selects?

And if that's the case (the file needs to be writeable by others), then if it's not writeable, it's a problem that really should be fixed and we can error out on that (in the business logic code itself, not here), can't we?

Reply:
Right, I tried to save one code branch, but that did not work.
I changed the code once more.
Root must try an update because database and journal(s) must be writable.
Unprivileged user must try s select, read-only is sufficient.